test/e2e/metrics_test.go

```diff
@@ -89,134 +129,265 @@ func NewMetricsTestConfig(t *testing.T, client, selector, clusterRole, clusterBi
 	}
 }
 
-// run will execute all steps of those tests
+// run executes the entire test flow
 func (c *MetricsTestConfig) run() {
-	c.createMetricsClusterRoleBinding()
-	token := c.getServiceAccountToken()
-	c.createCurlMetricsPod()
-	c.validate(token)
-	defer c.cleanup()
-}
-
-// createMetricsClusterRoleBinding to binding and expose the metrics
-func (c *MetricsTestConfig) createMetricsClusterRoleBinding() {
-	c.t.Logf("Creating ClusterRoleBinding %s in namespace %s", c.clusterBinding, c.namespace)
-	cmd := exec.Command(c.client, "create", "clusterrolebinding", c.clusterBinding,
-		"--clusterrole="+c.clusterRole,
-		"--serviceaccount="+c.namespace+":"+c.serviceAccount)
-	output, err := cmd.CombinedOutput()
-	require.NoError(c.t, err, "Error creating ClusterRoleBinding: %s", string(output))
-}
-
-// getServiceAccountToken return the token requires to have access to the metrics
-func (c *MetricsTestConfig) getServiceAccountToken() string {
-	c.t.Logf("Generating ServiceAccount token at namespace %s", c.namespace)
-	cmd := exec.Command(c.client, "create", "token", c.serviceAccount, "-n", c.namespace)
-	tokenOutput, tokenCombinedOutput, err := stdoutAndCombined(cmd)
-	require.NoError(c.t, err, "Error creating token: %s", string(tokenCombinedOutput))
-	return string(bytes.TrimSpace(tokenOutput))
-}
-
-// createCurlMetricsPod creates the Pod with curl image to allow check if the metrics are working
-func (c *MetricsTestConfig) createCurlMetricsPod() {
+	ctx := context.Background()
+	c.createMetricsClusterRoleBinding(ctx)
+	token := c.getServiceAccountToken(ctx)
+	c.createCurlMetricsPod(ctx)
+	c.waitForPodReady(ctx)
+	// Exec `curl` in the Pod to validate the metrics
+	c.validateMetricsEndpoint(ctx, token)
+	defer c.cleanup(ctx)
```

Comment:
nit: I know that wasn't introduced by your PR, but `defer` looks redundant, as it's already being called as the last call

Reply:
Yeah I think it inadvertently got moved down. Moving back to the top of the function since point was to cleanup no matter what (pretty sure)

Reply:
I agree, good point